### PR TITLE
Detect overflow in proc_macro_server subspan

### LINKED ISSUE #72237: Integer overflow on String::drain() with an inclusive range
I would expect this code to panic at the `drain()` because `usize::max_value()` is an out of bounds index:

```rust
fn string_drain_overflow() {
    let mut string = String::new();
    string.drain(..=usize::max_value());
}
```

Instead, it succeeds and returns an empty iterator. Looking at the [implementation for `String::drain()`](https://doc.rust-lang.org/src/alloc/string.rs.html#1542), I'm assuming an integer overflow happens at line 1542 `Included(&n) => n + 1` and the range is taken to be `0..0`. Honestly, I expected the add operation itself to panic on overflow so I'm not entirely sure what the correct behaviour is here.

## Meta

```
rustc 1.43.1 (8d69840ab 2020-05-04)
binary: rustc
commit-hash: 8d69840ab92ea7f4d323420088dd8c9775f180cd
commit-date: 2020-05-04
host: x86_64-unknown-linux-gnu
release: 1.43.1
LLVM version: 9.0
```

Update: the same bug exists in [`String::replace_range`](https://doc.rust-lang.org/src/alloc/string.rs.html#1592).

### PR DESCRIPTION
**Pull request by tmiasko**
* Detect overflow in proc_macro_server subspan
* Add tests for overflow in Vec::drain
* Add tests for overflow in String / VecDeque operations using ranges

### CODE REVIEW

File: library/alloc/src/collections/vec_deque.rs [hunk 1]
```diff
@@ -1090,11 +1090,13 @@ impl<T> VecDeque<T> {
         let len = self.len();
         let start = match range.start_bound() {
             Included(&n) => n,
-            Excluded(&n) => n + 1,
+            // Detect overflow with bounds checking code below.
+            // Using a saturated index is fine, because usize::MAX index is never valid.
```
Comment (sfackler) [inline library/alloc/src/collections/vec_deque.rs:1094]: I think it can be valid for a deque of ZSTs?

Comment (tmiasko) [reply library/alloc/src/collections/vec_deque.rs:1094]: Currently this is not possible and I think unlikely to change. The capacity is a power of two (for zero sized types MAXIMUM_ZST_CAPACITY is used) so maximum size is much smaller than that. Additionally one element is always left vacant.

**Comment by Mark-Simulacrum** [comment]
@estebank -- I note that the issue suggested that we were unsure about the DWIM semantics you meant (https://github.com/rust-lang/rust/issues/72237#issuecomment-665706873).

That said this has been open for a month now and looks reasonable to me; if there are semantics that move us from panicking to not panicking I think that'd relatively (if not perfectly) backwards compatible to move to. I am going to approve it, but we can revert or otherwise pause if necessary.

@bors r+

**Comment by tmiasko** [comment]
The #75207 introduced a new bounds checking API for ranges, and used it to fix
most of overflow issues, making most of changes here unnecessary. It didn't add
any new tests, so I retained those.

**Pull request merged**